File: ios/LaneShadow/Views/Atoms/Badge.swift
import LaneShadowTheme
import SwiftUI

/// Badge component - Small label or tag with semantic theme styling
///
/// Following the translation matrix specification:
/// - paddingHorizontal: 10 (custom value between sm and md tokens)
/// - paddingVertical: 2 (half of xs token)
/// - borderRadius: full (9999 for pill shape)
/// - Typography: type.label.sm (12sp / 18lh / 600w)
public struct Badge: View {

    // MARK: - Properties

    @Environment(\.theme) private var theme

    private let text: String
    private let variant: BadgeVariant
    private let icon: String?
    private let opacity: Double

    // MARK: - Initialization

    /// Creates a Badge with the given text and optional styling
    /// - Parameters:
    ///   - text: The text to display
    ///   - variant: The color variant (default is .default)
    ///   - icon: Optional icon name to display
    ///   - opacity: Opacity value (default is 1.0)
    public init(
        _ text: String,
        variant: BadgeVariant = .default,
        icon: String? = nil,
        opacity: Double = 1.0
    ) {
        self.text = text
        self.variant = variant
        self.icon = icon
        self.opacity = opacity
    }

    // MARK: - Body

    public var body: some View {
        HStack(spacing: theme.space.xs) {
            if let icon {
                Image(systemName: icon)
                    .font(.system(size: 12))
                    .foregroundStyle(resolvedColors.foreground)
            }

            Text(text)
                .font(.system(size: theme.type.label.sm.fontSize, weight: .semibold))
                .foregroundStyle(resolvedColors.foreground)
        }
        .padding(.horizontal, 10) // Matrix: paddingHorizontal = 10
        .padding(.vertical, 2)   // Matrix: paddingVertical = 2
        .background(resolvedColors.background)
        .clipShape(Capsule(style: .continuous)) // Matrix: borderRadius = full
        .overlay {
            // Border for outline variant
            if resolvedColors.border != nil {
                Capsule(style: .continuous)
                    .stroke(resolvedColors.border!, lineWidth: 1) // Matrix: borderWidth = 1
            }
        }
        .opacity(opacity) // Matrix: opacity prop support
        .accessibilityElement(children: .combine)
        .accessibilityLabel(text)
    }

    // MARK: - Private Helpers

    private var resolvedColors: BadgeResolvedColors {
        BadgeResolvedColors.resolve(theme: theme, variant: variant)
    }
}

// MARK: - BadgeVariant

/// Badge color variants matching the translation matrix
public enum BadgeVariant: String, CaseIterable {
    case `default`
    case secondary
    case destructive
    case outline
    case success
    case warning
    case info
}

// MARK: - BadgeResolvedColors

private struct BadgeResolvedColors {
    let background: Color
    let foreground: Color
    let border: Color?

    static func resolve(theme: Theme, variant: BadgeVariant) -> Self {
        switch variant {
        case .default:
            return Self(
                background: theme.colors.primary.default,
                foreground: theme.colors.onPrimary.default,
                border: nil
            )
        case .secondary:
            return Self(
                background: theme.colors.secondary.default,
                foreground: theme.colors.onSecondary.default,
                border: nil
            )
        case .destructive:
            return Self(
                background: theme.colors.danger.default,
                foreground: theme.colors.onPrimary.default,
                border: nil
            )
        case .outline:
            return Self(
                background: .clear,
                foreground: theme.colors.onSurface.default,
                border: theme.colors.border.default
            )
        case .success:
            return Self(
                background: theme.colors.success.default,
                foreground: theme.colors.onPrimary.default,
                border: nil
            )
        case .warning:
            return Self(
                background: theme.colors.warning.default,
                foreground: theme.colors.onPrimary.default,
                border: nil
            )
        case .info:
            return Self(
                background: theme.colors.info.default,
                foreground: theme.colors.onPrimary.default,
                border: nil
            )
        }
    }
}

// MARK: - Preview

#Preview("Default") {
    Badge("Default")
        .laneShadowTheme()
}

#Preview("Secondary") {
    Badge("Secondary", variant: .secondary)
        .laneShadowTheme()
}

#Preview("Destructive") {
    Badge("Destructive", variant: .destructive)
        .laneShadowTheme()
}

#Preview("Outline") {
    Badge("Outline", variant: .outline)
        .laneShadowTheme()
}

#Preview("Success") {
    Badge("Success", variant: .success)
        .laneShadowTheme()
}

#Preview("Warning") {
    Badge("Warning", variant: .warning)
        .laneShadowTheme()
}

#Preview("Info") {
    Badge("Info", variant: .info)
        .laneShadowTheme()
}

#Preview("With Icon") {
    Badge("New", variant: .success, icon: "checkmark.circle.fill")
        .laneShadowTheme()
}

#Preview("With Opacity") {
    VStack(spacing: 8) {
        Badge("100% opacity", variant: .default, opacity: 1.0)
        Badge("75% opacity", variant: .default, opacity: 0.75)
        Badge("50% opacity", variant: .default, opacity: 0.5)
        Badge("25% opacity", variant: .default, opacity: 0.25)
    }
    .laneShadowTheme()
}

File: android/app/src/main/java/com/laneshadow/ui/components/atoms/Badge.kt
package com.laneshadow.ui.components.atoms

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import com.laneshadow.theme.LocalLaneShadowTheme

/**
 * Badge variant types
 *
 * Following RN wrapper API from react-native/components/ui/badge.tsx
 */
enum class BadgeVariant {
    Default,
    Secondary,
    Destructive,
    Outline,
    Success,
    Warning,
    Info,
}

/**
 * Badge component props
 *
 * Following RN wrapper API from react-native/components/ui/badge.tsx
 *
 * @param variant Badge color variant (default, secondary, destructive, outline, success, warning, info)
 * @param text Text content to display
 * @param icon Optional icon composable to display before text
 * @param opacity Opacity value for semi-transparent backgrounds (default: 1.0)
 * @param modifier Modifier for the container
 */
@Composable
fun Badge(
    variant: BadgeVariant = BadgeVariant.Default,
    text: String,
    icon: @Composable (() -> Unit)? = null,
    opacity: Float = 1.0f,
    modifier: Modifier = Modifier,
) {
    val theme = LocalLaneShadowTheme.current

    // Get background color based on variant
    val backgroundColor: Color = when (variant) {
        BadgeVariant.Secondary -> theme.colors.secondary.default
        BadgeVariant.Destructive -> theme.colors.danger.default
        BadgeVariant.Success -> theme.colors.success.default
        BadgeVariant.Warning -> theme.colors.warning.default
        BadgeVariant.Info -> theme.colors.info.default
        BadgeVariant.Outline -> Color.Transparent
        BadgeVariant.Default -> theme.colors.primary.default
    }

    // Get text color based on variant
    val textColor: Color = when (variant) {
        BadgeVariant.Secondary -> theme.colors.onSecondary.default
        BadgeVariant.Outline -> theme.colors.onSurface.default
        else -> theme.colors.onPrimary.default
    }

    // Get border style for outline variant
    val border: BorderStroke? = when (variant) {
        BadgeVariant.Outline -> BorderStroke(
            width = 1.dp,
            color = theme.colors.border.default
        )
        else -> null
    }

    Surface(
        modifier = modifier.alpha(opacity),
        shape = androidx.compose.foundation.shape.CircleShape,
        color = backgroundColor,
        border = border,
    ) {
        Row(
            modifier = Modifier
                .padding(horizontal = 10.dp, vertical = 2.dp),
            horizontalArrangement = Arrangement.Center,
            verticalAlignment = Alignment.CenterVertically,
        ) {
            if (icon != null) {
                Row(
                    modifier = Modifier.padding(end = theme.space.xs),
                    horizontalArrangement = Arrangement.Center,
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    icon()
                }
            }
            Text(
                text = text,
                style = theme.type.label.sm,
                color = textColor,
            )
        }
    }
}
